### ISSUE TITLE
opening hours form field

### ISSUE THREAD

**Issue by ansis** [openstreetmap/iD#974]
http://wiki.openstreetmap.org/wiki/Key:opening_hours

**Comment by M1dgard** [comment]
Any progress here?
In the mean time maybe a link could be provided to simple instructions (a [page like this](https://wiki.openstreetmap.org/wiki/User:M!dgard/How_to_map_opening_hours)) for the syntax. Many contributors just fill in something in natural language, which is useless to automated interpreters like [OpenLinkMap](http://openlinkmap.org) and [OsmAnd](http://osmand.net).

**Comment by bagage** [comment]
Also another great alternative not mentioned yet is [YoHours](http://github.pavie.info/yohours/). 

Maybe iD could give an hint to use one of these solutions when starting writing opening hours manually? 
It would not be a hard solution to implement right?

**Comment by tmcw** [comment]
> It would not be a hard solution to implement right?

Please implement it, if you perceive it to be easy.

**Comment by bhousel** [comment]
> Also another great alternative not mentioned yet is YoHours.
> 
> Maybe iD could give an hint to use one of these solutions when starting writing opening hours manually? 

I just spent about 10 minutes playing around with it and it's unfortunately kind of buggy and hard to use.  I do think it's probably a good starting effort though.  If it were a bit more polished and could look nice in a lightboxed small dimension iframe, I'd consider it.

> It would not be a hard solution to implement right?

I'm vetoing what @tmcw says above, this is too buggy to use currently and I wouldnt merge it.  For now spend time improving the tool, but not on integrating it with iD.

**Comment by iandees** [comment]
Also, supporting all the complex features behind the `opening_hours` tag language is nice, but the interface on YoHours is *way* too complex for iD. Limit to full (or half maybe) hours, don't expose the seasonal options, and only allow one tab to start. If someone wants to build more complex OH strings then they could switch to some advanced mode or something.

**Comment by bagage** [comment]
@tmcw It was more of a question than an assumption, I would certainly not state on unknown technologies; my apologies if it looked like it. 
I just wanted to bump the issue since I found `YoHours` a bit earlier and found it valuable to map a swimming pool with seasonal opening hours which would have take me half an hour to do manually. 
I am not `YoHours` author at all, but considered it as easy enough to use in iD. It seems I was wrong :). At least it may help on what-to-do and what-not-to-do for iD widget.

**Comment by tyrasd** [comment]
//cc @PanierAvide: would you like to comment on the issues raised by @bhousel above (https://github.com/openstreetmap/iD/issues/974#issuecomment-133526716)?

**Comment by PanierAvide** [comment]
Of course, so I'm the author of this buggy interface :) For the while it's not meant to be included in something else, but I was thinking of providing a simple HTML page which could be embed in an iframe. What would be the requirements to include this iframe in iD (dimensions, functionalities, show text input or not, how to expose result value...) ?

**Comment by PanierAvide** [comment]
I made a first attempt of a [minimal embeddable version of YoHours](http://github.pavie.info/yohours/iframe.html?oh=We%2001:30-12:45). Things can still be smaller, and the seasons dialog content should be optimized. Let me know if it has any interest.

**Comment by d1g** [comment]
@PanierAvide  I will be happy to see yohours in iD.

Normally we want time table and columns with 24/7 and ("Monday"...) headers, Other settings are less popular and should be displayed as small tabs.

We may want to add more *hover* explanations to table in user language about how advanced features of opening_hours work. yohours contain no context-sensitive help or hover captions.

**Comment by PanierAvide** [comment]
Is there any hover caption system specific to iD, or something as simple as `<a title="...">` can be used ?

**Comment by d1g** [comment]
@PanierAvide I think it is ID-specific, You can test it at way with highway=track. CSS class is "form-field-tracktype". 

Unlike html select it will:
1. display shorter/trimmed title
2. provide user with translated string

We will need hovers caption like this, but for calendar controls (sliders) or text fields (hover on opening_hours=24/7 will show "Open 24/7" in user language)

PS. We can annotate most OH strings in natural language, but it will be more demanding task.

**Comment by rugk** [comment]
Any progress here? I think `opening_hours` is one of the most difficult days a Newbie encounter.

So iD should:
1. Maybe at least validate the entered value, to prevent bad values
2. (Later) offer a nice UI for editing it

Note that also [Maps.me](https://github.com/mapsme) has a simple `opening_hours` editor, which works great for most cases.

**Comment by boothym** [comment]
> Any progress here? I think opening_hours is one of the most difficult days a Newbie encounter.

Indeed and as a newbie you probably aren't going to look into the wiki and see exactly how it's done. One I saw today from a new user was "Mon-Fri 0800 - 18.30, Sat 0800-1700, Sun 0800-1400" - a shame it was so close but not correct. I wonder if iD could do some basic validation, like checking for two letter days and time separator?

@bhousel here's a couple of examples:
### google mapmaker desktop
Clicking on Mon-Sat gives you a dropdown which can be ticked, the hours also give a dropdown with half hourly times.
![hours](https://cloud.githubusercontent.com/assets/16965772/22040237/14c9f832-dcfa-11e6-9c89-c0a59ee6abb4.jpg)
### maps.me android
The advanced mode gives the normal alphanumeric field and some examples of the opening_hours format.
![maps](https://cloud.githubusercontent.com/assets/16965772/22040236/14a67268-dcfa-11e6-9fc8-cfa83c30adda.png)

**Comment by bhousel** [comment]
To everyone commenting on this ticket.  I don't need inspiration on how to design a date picker.  The reason this field has not been built yet is not because I don't know what a date picker is, or know where to find one.

The problem here is that `opening_hours` is just a terribly complex grammar and iD needs to support most of it.  For example, `sunrise-sunset` is [currently the 8th most common value](http://taginfo.openstreetmap.org/keys/opening_hours#values) for `opening_hours`.  If somebody edits a business already tagged with `opening_hours=sunrise-sunset`, what should this field show?  We try very hard not to break existing data.

I do want to support this, but it's not as simple as dropping in a date picker.

**Comment by iandees** [comment]
What if a day/time picker is only shown when the value of `opening_hours` matched a regex that covers the "simple case"? e.g. `Mo-Fr 09:00-17:00; Sa-Su 10:00-13:00`. If it doesn't match such a regex, it shows the standard text box?

**Comment by boothym** [comment]
Sorry, didn't realise where you were at with this issue :P

In addition to finding the simple case, you could also find the "complex" cases and only show the text box. For example, if it contains dawn | sunrise | sunset | dusk | unknown | open | closed | off, anything with "", etc.

**Comment by rugk** [comment]
I agree to @iandees: Just try to parse the existing field, if it fails fall back to the regular input box (and probably offer a button: "switch to easy input mode")

This would be very similar as in Maps.me, which also has a really basic input date/time picker: At the bottom of the screen you can see "advanced mode", which offers the traditional input box. However the time can be overwritten by the visual time picker.

**Comment by boothym** [comment]
@bhousel any progress with this issue? Reason I ask is I've seen a number of new users add things like "11am to Midnight", "10:30am - 9:30pm" or spelling out the days of the week.

Even if it's not a day/time picker at the moment, there should be something to help new mappers add opening hours in the correct format, especially in a blank text box. How about a placeholder with "Mo-Fr 09:00-17:00; Sa 10:00-14:00" or similar?

**Comment by bhousel** [comment]
For now, I could make the field a dropdown, then users would at least get some valid things suggested to them.

**Comment by rugk** [comment]
Yeah, offering the "most common" values or so could be a good start.

**Comment by bhousel** [comment]
> For now, I could make the field a dropdown, then users would at least get some valid things suggested to them.

Ok this was done.  CCing @joto because `opening_hours` is one of those fields with a lot of values (see #3955), but not as bad as `name` or `ref`.  So we should watch that it doesn't affect taginfo performance too negatively.

**Comment by 1ec5** [comment]
> The problem here is that `opening_hours` is just a terribly complex grammar and iD needs to support most of it.

Here are the constraints for this feature as I see it:

* The `opening_hours` syntax is very complex.
* A UI for inputting the value needs to be intuitive without requiring too much form-filling (not too fiddly).
* The input UI also needs to fit within the feature editor, which is constrained horizontally.
* The output UI needs to be understandable at a glance – the mapper needs to be able to scan it quickly and know whether it’s the correct value.

Many calendaring applications have a compact, unfiddly UI for entering new event details: just a freeform textbox that the application parses when you hit OK. The best implementations treat the input as natural language instead of requiring the user to conform the input to a particular format. If this is a suitable input mechanism, then I think a prepopulated text field would also be a suitable output mechanism as well. Imagine seeing `Weekdays from 9 AM to 5 PM`, correcting it to `Monday-Thursday from 9 AM to 5 PM`, and having it autocorrect to `Mondays through Thursdays from 9 AM to 5 PM` with `Mo-Fr 09:00-17:00` as the underlying tag value.

#3269 suggested using the open-source [hoursparser.js](https://github.com/foursquare/hoursparser.js) library to implement simple opening-hours input. This library is limited to English and lacks any output formatting functionality.

Another library, [rrule](https://github.com/jakubroztocil/rrule), specializes in working with the RRule recurring event rule format found in [iCalendar](https://en.wikipedia.org/wiki/ICalendar) events. This library supports [converting both ways](https://github.com/jakubroztocil/rrule#natural-language-text-methods) between RRule and natural language. This library is nominally internationalized, though it lacks translations and probably needs more work to handle languages that differ grammatically from English: jakubroztocil/rrule#13.

I wonder if it would be feasible to convert between the `opening_hours` syntax and RRule in order to take advantage of the rrule library. Or maybe we could extract the NLP portions of rrule and adapt them to `opening_hours`.

Edit: [This tool](http://lists.openstreetmap.ch/pipermail/talk-ch/2018-December/009832.html) ([repo](https://gitlab.com/geometalab/web-to-osm-opening-hours/blob/master/README.md)) converts from German to `opening_hours`. It could also be a model for something in iD.

**Comment by iandees** [comment]
We shouldn't make this too complicated. We should copy the websites/apps that have already figured this out.

For example, Google's looks like this:

![image](https://user-images.githubusercontent.com/261584/48093124-64c05a00-e1c3-11e8-92bd-3ad0cb1b8ecf.png)

![image](https://user-images.githubusercontent.com/261584/48093137-6ee25880-e1c3-11e8-8fb9-68acc0aece5a.png)

This covers the vast majority of simple cases and [as mentioned earlier](https://github.com/openstreetmap/iD/issues/974#issuecomment-273379916) we could skip trying to parse more complicated cases.

**Comment by rugk** [comment]
Exactly, there are many websites/apps you could take inspiration from: Maps.me, StreetComplete, any calendar app, …

**Comment by skquinn** [comment]
At the very least, data for opening_hours, service_times, and the like needs to be validated by iD, even if there's no full-blown editor that can handle every edge case of the syntax. I've cleaned up a bunch of them over the past couple of days (including the entire states of Rhode Island, Wyoming, North Dakota, South Dakota, the majority of Kansas, and sizeable chunks of Texas) but without something telling the user "Hey, this syntax is wrong, please fix" they will keep slopping down more rubbish making the tag nearly useless.

**Comment by TheAdventurer64** [comment]
Any progress?

**Comment by quincylvania** [comment]
@TheAdventurer64 Any progress would be posted here.

**Comment by ghost** [comment]
Looks like it's been a few years since YOHours was mentioned in the thread, are the usability concerns still present? If the situation has improved, even something like the attached mockup (except adjusted for iD's style, I just faked it quickly in dev tools) would be an improvement / useful stopgap -- it seems many users don't concern themselves with the raw tags section, only the widgets area.
![image](https://user-images.githubusercontent.com/634501/139605766-94a74fba-7690-4390-beb6-56aed856d440.png)

**Comment by rmikke** [comment]
Someone has alredy coded part of JOSM-like opening hours box:
https://github.com/Deuxis/timesetter

**Comment by tordans** [comment]
> Someone has already coded part of JOSM-like opening hours box…

There is also a working version in MapComplete.
* Example: https://pietervdvn.github.io/mc/develop/cyclofix.html?z=17&lat=52.4752&lon=13.4441&language=en#node/1954982655
* Docs for how to add it to a theme https://github.com/pietervdvn/MapComplete/blob/8ea6f494bf26953945b38fddf2fc60f7d98aaa2a/Docs/SpecialRenderings.md#opening_hours_table
* Code https://github.com/pietervdvn/MapComplete/blob/master/UI/OpeningHours/OpeningHoursVisualization.ts#L15

This might be an interface component, that multiple JS based OSM editors can use; so building this as a separate component sounds like an interesting idea.

**Comment by Mashin6** [comment]
It would be really helpful if there was some movement on this issue. There are lot of people adding opening hours to restaurants around them and it takes time to fix it after them. 

What about adding support for very simple cases. Like for example the time picker would show only if the `opening_hours` field is empty or parsable without error (to avoid problems with sunset-sunrise etc.) and it would be only for one time slot per day. No months, no holidays.

It could be as simple as 7 sliders that could be rolled out as a drop down menu.

![image](https://user-images.githubusercontent.com/5265707/142332989-03032b8d-b97d-4f96-abc2-3588b0af5a34.png)

**Comment by iandees** [comment]
There's been a lot of really great screenshots of others doing this. Someone just has to spend the time to make the change to iD 😄

**Comment by lefuturiste** [comment]
Hello, just saying that I'm currently working on a PR to improve iD support of opening hours.
I think that the first step would be to validate the `opening_hours`, `collection_times` and `service_times` fields using [opening_hours.js](https://github.com/opening-hours/opening_hours.js). The next step is to have a basic editor that cover basic needs.

**Comment by 1ec5** [comment]
> Looks like it's been a few years since YOHours was mentioned in the thread, are the usability concerns still present? If the situation has improved, even something like the attached mockup (except adjusted for iD's style, I just faked it quickly in dev tools) would be an improvement / useful stopgap -- it seems many users don't concern themselves with the raw tags section, only the widgets area.

The earlier concerns about YoHours didn’t come with specific details or acceptance criteria, so I’m also willing to assume that improved access to the tool would be a net positive for most users. #9678 implements something similar to the mockup in https://github.com/openstreetmap/iD/issues/974#issuecomment-955817015 as a stopgap.

> I think that the first step would be to validate the `opening_hours`, `collection_times` and `service_times` fields using [opening_hours.js](https://github.com/opening-hours/opening_hours.js). The next step is to have a basic editor that cover basic needs.

Once we have at least a minimal UI like the one in #9678, validation would be a sensible next step, but not before then. It wouldn’t be fair to warn users about broken syntax without offering a solution. opening_hours.js does have validation functionality we could use, though it might be overkill compared to generating a tokenizer just for the purpose of validation. For example, I whipped up [a formal definition](https://gist.github.com/1ec5/a8aa66b6b286783949aa18918f513327) of the opening_hours syntax for [Peggy](https://peggyjs.org/); though it definitely has some kinks to work out, it would get the job done without needing to analyze the value for actual date ranges.

Beyond that, I would reiterate the point in https://github.com/openstreetmap/iD/issues/974#issuecomment-436404516 that all the calendar mockups so far probably fall short of what iD’s inspector needs. A calendar interface is undoubtedly useful for editing, but it takes up far too much space to appear by default in every preset, including when the user merely hovers over a feature to preview its tags. We wouldn’t be able to hide the raw syntax, which is kind of how this all started. There needs to be a more compact representation, similar to the one opening_hours.js’s `prettifyValue()` function returns, but localized. Map applications typically display a similar value as a subtitle in a POI detail view.

**Comment by wcedmisten** [comment]
Hello! There's been a lot of discussion on this over the last 10 years. My thoughts are that @Mashin6's suggestions of using sliders may one of the easier approaches for a MVP supporting the simplest/most common subset of opening_hours syntax.

I took a first stab at implementing this, although it's very far from being usable and the CSS is still quite ugly, I'd like to get feedback on it before going further in this direction.

![image](https://github.com/openstreetmap/iD/assets/4565200/f9798df8-3018-47dc-8a9c-94a82d2b8c32)

[This comment](https://github.com/openstreetmap/iD/issues/974#issuecomment-273378730) makes me think the maintainers would prefer a more comprehensive solution, rather than merge in something that only supports of a subset of opening_hours.

My personal opinion is that we should provide a good UI for the simpler syntax and just suggest manual editing of tags for the complex syntax. Linking to another tool like https://github.com/openstreetmap/iD/pull/9678 to support the complex syntax also seems like a viable workaround.

Would any of the maintainers care to weigh in with their present opinions on this to inform my future work? Thanks in advance!

**Comment by 1ec5** [comment]
Thanks for taking the time to make this mockup. I’m not the official maintainer, but perhaps I could elaborate on my contrarian comments so far.

Many of the mockups so far are optimized for the simple kind of opening hours that has a single timespan each day of the week, consistent across all weeks of all years. I agree that this is one of the most important cases to address (with varying levels of importance depending on the POI type and region). It’s fair to start with the most important cases and figure out how to shoehorn edge cases into the UI later. Your proposal would actually make #9678 completely unnecessary, because it covers the same exact cases.

However, I think most users who need to use this field are copying opening hours from a sign or website or maybe from memory. These opening hours are always expressed as literal times, so that’s what the user would be able to enter most readily. Any UI that requires dragging, whether on a calendar or a slider, introduces a degree of imprecision. Now you have to watch the time indicator while you drag the slider around. If the shop opens at a quarter past the hour, you’ll need good hand-eye coordination. Where I come from, many schools open or close at five or eight minutes past the hour for some reason.

Pairing each double-knobbed slider with two editable text fields would allow the user to adjust the value to match the sign. But at that point, I wonder if the slider is providing any value compared to just the two text fields (perhaps with dropdown menus or increment/decrement buttons). https://github.com/openstreetmap/iD/issues/974#issuecomment-436405783 has an example of what a text field–based design could look like.

There’s still the issue of minimizing the field’s vertical space. Maybe the field could start out in its current form, as a single unstructured text field, but when you focus the field, a more structured form could slide out below. This wouldn’t quite prevent the user from ever seeing the raw opening hours syntax, but at least it would get the main editing case out of the way so we can focus on a more comprehensive solution.

**Comment by wcedmisten** [comment]
Yeah that makes sense, now that I think about it, the sliders seem a bit fiddly for entering hours, even if limited to 1-hour intervals.

With https://github.com/openstreetmap/iD/issues/974#issuecomment-436405783, does "add hours" open a separate block to enter other days? Otherwise, how would the user enter e.g. separate weekend hours with that interface?

Or is that not supported in that UI?

**Comment by 1ec5** [comment]
I assume that screenshot was from the Google Maps UI for suggesting changes. These days, the UI takes up a lot more space; it’s more like a wizard than an inspector now. Clicking “Add hours” lets you add another time span for the selected day or days.

<img src="https://github.com/openstreetmap/iD/assets/1231218/2d77c6d7-335b-42f8-8b01-8406b1b24abe" width="224" alt="Hours"> <img src="https://github.com/openstreetmap/iD/assets/1231218/8d911f06-f54e-4875-a366-073b47165f11" width="280" alt="Edit all hours">

**Comment by LySioS** [comment]
It's been 11 years now. 

This basic need is still being ignored... 

Can we solved it ? @tyrasd

**Comment by LySioS** [comment]
Up 😔

**Comment by Dimitar5555** [comment]
Here is another design for inspiration, taken from Foursquare.com.

On the left you choose day(s) and enter hours, on the right the current data is shown.

<img width="1328" height="291" alt="Image" src="https://github.com/user-attachments/assets/1087b348-da6c-41c8-b7a3-bef379e20417" />

Here is the form for seasonal hours (the two checkboxes don't seem to work):

<img width="1328" height="430" alt="Image" src="https://github.com/user-attachments/assets/8dead02c-d37d-4bae-8e65-011d57d065bb" />

You can't edit them, but from a UX POV, it should be possible to click any hour, have its data loaded and edited.

Our opening hours schema is a bit more complex so for anything that can't be recognised easily, it should be shown as text.
